lib/isqrt.man
.\" RCSID @(#)$Id: isqrt.man,v 1.1 2002/07/18 22:11:15 rkowen Exp $
.\" LIBDIR
.TH "ISQRT" "3rko" "26 Nov 1998"
.SH NAME
isqrt \- integer sqrt
.SH SYNOPSIS

.nf
#include "isqrt.h"

         char  chsqrt(char n);
  signed char  scsqrt(signed char n);
unsigned char  ucsqrt(unsigned char n);
         short  hsqrt(short n);
unsigned short uhsqrt(unsigned short n);
         int    isqrt(int n);
unsigned int   uisqrt(unsigned int n);
         long   lsqrt(long n);
unsigned long  ulsqrt(unsigned long n);
.fi

.SH DESCRIPTION
The
.I isqrt
collection returns the nearest lower integer square root for the given
integer argument.  It uses a method similar and augmenting the one
described by Arne Steinarson (arst@ludd.luth.se).
(see http://www.aha.ru/~pervago/articles/sqroot.txt or
http://www.df.lth.se/~john_e/gems/gem0034.html)
.P
It finds the lowest even power of two (2^(2n)) that contains the argument,
shifting the argument to the right by 2*n bits (which corresponds to
division by 2^(2n)).  This leaves a value between [0,1), which is
used to search a lookup table (LUT).  The table is accurate to 8 bits.
Once the value is found in the LUT, the value is left shifted by n bits,
which corresponds to multiplication by 2^n.
Finally, the Newton method to quadratically converge to a more accurate
value.
.P
Here's roughly an example of what's going on.  Take the isqrt(10000).
10000 = 0x2710, which is less than 2^14 but greater 2^12.  Hence shift
the value 14 bits to the right which yields .610352.  The square root
is computed (.78125) which is then shifted to the left by 7 bits, which
results in 100, as expected!
Of course, we use a LUT, which only gets the value approximately right.
The exact value is generated by using the Newton method, which quadratically
converges the zeros of functions by using the current value and slope
generated from the function.
.P
This method is extremely fast and involves no conversion to any
floating point representations and std-C math libraries.
The code makes no assumptions about
the length of any of the larger integer types except that they are
larger than an unsigned char.
However, the code can be made even faster by using a binary search to
determine 2^(2*n) if the length of the integer type is known.

.SH SEE ALSO
sqrt(3),gcd(3rko)

.SH DIAGNOSTICS
Returns 0 for invalid input values.

.SH AUTHOR
R.K.Owen,Ph.D.

.KEY WORDS


NEWS
* Note: Implied Consent -
  If you send in a code patch or fix of one type or another you consent
  to have your name and email address listed here unless you explicitly
  note otherwise.  Whether your name & email is actually listed depends
  on how extensive the change.  (If it's a minor code, document change,
  or spelling correction your name will probably not be listed.)

****************************   Release 3.2 ******************************
Modules 3.2.9
	* Fix the coexistence of load/remove flags (Martin Siegert)
	* Fixed an overlapping string copy (Wilson Snyder & Orion Poplawski)
	* Isolated calls to the new interpreter and get & set env.vars.
	* Optionally use Tcl memory check routines for malloc/realloc.
	* Fixed the "module purge" memory corruption (Poor Yorick)
	* Added support for Ruby (Tammo Tjarks)

Modules 3.2.8
Oct  1, 2010				R.K. Owen (rk@owen.sj.ca.us)
	* Handle the .modulerc/.version in deep modulefile dirs,
	  Patches thanks to Lindsay Todd, and others.
	* Note that .version takes precedence over .modulerc files.
	* Renamed init/python to init/python.py and updated with
	  anonymous patch to eliminate 'sed' references
	* Renamed init/perl to init/perl.pm
	* Added brief section to module.1 on invoking modules for perl & python
	* Added Nathaniel Waisbrot's patch for CMake support with tests.
	* Fixed the MANPATH in modulefiles/modules to use configure generated
	  values.
	* Added the configure --with-man-path so a system specific MANPATH
	  could be used.  Some "man"s radically change their search path
	  when a MANPATH exists and when not.
	  The default is "/usr/man:/usr/share/man" (when both exists).
	* Purging a purged module list does not result in an error.
	* Fixed the configure script for searching for tclConfig.sh
	  and handle --without-x in the tests.

Modules 3.2.7
Jul 30, 2009				R.K. Owen (rk@owen.sj.ca.us)
	* Have init/bash export the module function (thanks to Roy Dragseth)
	* Fixed swap bug when prepending or appending multiple times
	  (fix thanks to Gavin Walker), added tests for same.
	* Have purge remove loaded modules in reverse order
	  (patch contributed by carriees)
	* Added a batch of patches from Gerrit Renker
	  - Goes through each of the two manpages, bringing them up to date.
	  - Fix an error in configure.ac caused by misspelling
	  - Fix typos in help messages and source-code comments
	  - Fix the setup of $MODULESPATH to not have a trailing ':'
	  - Fix segmentation violation occurring when the shelltype is absent
	  - Fix a condition where 'make uninstall' does not cleanse a directory
	  - Fix a build problem which set the executable bit on 'data' files
	  - Cleans up noise in '.modulespath' file
	  - Replaces external sed filter script with internal Makefile.am rule
	  - Adds missing 'zsh' case in etc/global
	  - simplify (t)csh initialisation
	  - Add bash auto-completion for the 'module' command
	  - Adds new 'chdir' modulefile command (sets cwd on module load)
	* versions a & b fixed some bugs

Modules 3.2.6
Oct 29, 2007				R.K. Owen (rk@owen.sj.ca.us)
	* Fixed the locate_module.c:realloc() calls which didn't properly
	  take the element size into account (Fix thanks to Ron Isaacson).
Modules 3.2.5
Feb 14, 2007				R.K. Owen (rk@owen.sj.ca.us)
	* Wrapped all malloc() calls to make sure the size is never 0.
	  This fixes a problem spotted by Francesco Nigro under AIX,
	  and diagnosed by G Lehnert.
	* Updated tests when domainname is unknown.
Modules 3.2.4
Feb  2, 2007				R.K. Owen (rk@owen.sj.ca.us)
	* Fixed bug where "empty" directory paths in the MODULEPATH env.var.
	  needed to be handled gracefully.
Modules 3.2.3
Jun  1, 2006				R.K. Owen (rk@owen.sj.ca.us)
	* Allow version names to be substrings of other version names
	  thanks to dgoldendias at users.sourceforge.net
	* Fixed compiling bug when --enable-log was specified.
	* Made some minor changes to the syslogging and tests so that
	  they all pass.
	* the set-alias command will for Bourne shell derivatives
	  substitute the safe ${1+"$@"} for $*.
	* Added the --delim option to prepend-path and append-path to
	  allow arbitrary delimiters instead of the colon ':'
	* Fixed build problem with regards to the init/* files
	  (solution due to Thomas Zeiser)
Modules 3.2.2
May  4, 2006				R.K. Owen (rk@owen.sj.ca.us)
	* Replaced all strtok() calls with an internal version.
	* Now can append/prepend and empty string '' or colons ':'
	* Added more --with-tclx-* options in case tclxConfig.sh
	  can not be found, but the TclX lib does exist.
	* Expanded the search path for tclConfig.sh.
	* Updated the TclX init call for 8.4
Modules 3.2.1
Feb 13, 2006				R.K. Owen (rk@owen.sj.ca.us)
	* Catch 'exit N' within modulefiles, use this as an exit return code.
	  Has the behavior of loading the modulefile up to the point of exit.
	* For bash, a non-zero exit will append a test that returns false.
	* Added tests for "break" and "exit N"
	* various other bug fixes
Modules 3.2.0
Jan 19, 2006				R.K. Owen (rk@owen.sj.ca.us)
	* default for cache is now "no"
	* Added version_test() to testsuite to make some of the
	  tests conditional on the values generated by --version.
	* Reworked every configure-based file
	  - the configure options help is formatted with AC_HELP_STRING
	  - uses automake to generate most of the Makefile.in
	  - renamed configure.in to configure.ac per autoconf
	  - Versioning is optional now with a configure option
	  - reworked the init scripts for versioning or not
	  - eliminated acconfig.h and put templates in configure.ac
	  - make install now honors the DESTDIR variable for package builds
	  - honors bindir for modulecmd location
	   (but the init scripts are only correct if no versioning)
	* Collected the usage info and now it can be displayed with
	  the --help or -H option (same as help)
	* --version now gives most of the optional configure parameters
	  for easy debugging of problems
	* Made the output deterministic (sorted) and reworked tests also

lib/isqrt/isqrt.c
static const char RCSID[]="@(#)$Id: isqrt.c,v 1.1 2002/07/18 22:11:15 rkowen Exp $";
static const char AUTHOR[]="@(#)isqrt 1.0 1998/11/26 R.K.Owen,Ph.D.";
/* isqrt.c -
 * finds the the integer sqrt of the given integer.
 * returns 0 if a user error (such as the argument is non-positive)
 */
/** ** Copyright *********************************************************** **
 ** 									     **
 ** Copyright 2001 by R.K.Owen,Ph.D.		                      	     **
 ** last known email: librko@kooz.sj.ca.us				     **
 **                   rk@owen.sj.ca.us					     **
 ** see LICENSE.LGPL, which must be provided, for details		     **
 ** 									     **
 ** ************************************************************************ **/

#include "isqrt/isqrt_.h"

/* just like templates - almost */
_ISQRT(isqrt, int)

